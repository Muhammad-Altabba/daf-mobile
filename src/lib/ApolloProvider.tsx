import React from 'react'
import { ActivityIndicator } from 'react-native'
import { ApolloProvider } from 'react-apollo'
import { ApolloProvider as ApolloHooksProvider } from '@apollo/react-hooks'
import { ApolloClient, Resolvers } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { from } from 'apollo-link'
import { SchemaLink } from 'apollo-link-schema'
import { makeExecutableSchema } from 'graphql-tools'
import { Container, Screen, Text } from '@kancha/kancha-ui'

import { core, dataStore, db, resolvers, typeDefs } from './setup'

export const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
})

const contextLink = new SchemaLink({
  schema,
  context: { core, dataStore },
})

const link = from([contextLink])

export const cache = new InMemoryCache({})

export const client = new ApolloClient({
  connectToDevTools: true,
  cache,
  link,
})

interface Props {}
interface State {
  isRunningMigrations: boolean
}

class CustomProvider extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      isRunningMigrations: true,
    }
  }

  async componentDidMount() {
    await db.initialize()
    await dataStore.initialize()
    await core.startServices()

    this.setState({ isRunningMigrations: false })
    await core.syncServices(await dataStore.latestMessageTimestamps())
  }

  render() {
    if (this.state.isRunningMigrations) {
      return (
        <Screen>
          <Container flex={1} alignItems={'center'} justifyContent={'center'}>
            <ActivityIndicator size={'large'} />
          </Container>
        </Screen>
      )
    } else {
      return (
        <ApolloProvider client={client}>
          <ApolloHooksProvider client={client}>
            {this.props.children}
          </ApolloHooksProvider>
        </ApolloProvider>
      )
    }
  }
}

export default CustomProvider
