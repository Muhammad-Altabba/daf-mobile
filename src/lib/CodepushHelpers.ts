import codePush, { DownloadProgress } from 'react-native-code-push'
import Log from './Log'

export const handleCodePushStatusChange = (status: codePush.SyncStatus) => {
  switch (status) {
    case codePush.SyncStatus.CHECKING_FOR_UPDATE:
      Log.info('Checking for updates.', 'Codepush')
      break
    case codePush.SyncStatus.DOWNLOADING_PACKAGE:
      Log.info('Downloading package.', 'Codepush')
      break
    case codePush.SyncStatus.INSTALLING_UPDATE:
      Log.info('Installing update.', 'Codepush')
      break
    case codePush.SyncStatus.UP_TO_DATE:
      Log.info('Up-to-date.', 'Codepush')
      break
    case codePush.SyncStatus.UPDATE_INSTALLED:
      Log.info('Update installed.', 'Codepush')
      break
  }
}
