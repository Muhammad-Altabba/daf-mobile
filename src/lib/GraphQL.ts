import { ApolloClient, Resolvers } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { from } from 'apollo-link'
import { HttpLink } from 'apollo-link-http'
import gql from 'graphql-tag'

// const httpLink = new HttpLink({uri: '/graphql'})

const link = from([
  // httpLink,
])

const resolvers: Resolvers = {
  Query: {},
  Mutation: {
    log: (_, { type, message, category }, context) => {
      const query = gql`
        query getLogs {
          logs @client {
            id
            timestamp
            type
            category
            message
          }
        }
      `

      const previous = context.cache.readQuery({ query })
      const timestamp = new Date().getTime() / 1000
      const logItem = {
        id: `LogMessage:${timestamp}`,
        timestamp,
        type,
        message,
        category,
        __typename: 'LogMessage',
      }
      const data = {
        logs: [logItem, ...previous.logs],
      }

      context.cache.writeQuery({ query, data })
      return logItem
    },
  },
}

export const cache = new InMemoryCache()
export const client = new ApolloClient({
  connectToDevTools: true,
  typeDefs: gql`
    type LogMessage {
      id: ID!
      timestamp: Int
      type: Int
      message: String
      category: String
    }
    extend type Query {
      logs: [LogMessage!]
    }
  `,
  cache,
  link,
  resolvers,
})

cache.writeData({
  data: {
    logs: [],
  },
})
