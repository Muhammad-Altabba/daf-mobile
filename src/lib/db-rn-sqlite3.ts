import { openDatabase } from 'react-native-sqlite-storage'
import { Types } from './serto-graph'

export class RnSqlite implements Types.DbDriver {
  private db: any

  initialize(): Promise<any> {
    let setDb = (db: any) => {
      this.db = db
    }
    setDb = setDb.bind(this)

    return new Promise((resolve, reject) => {
      const db = openDatabase(
        { name: 'test', location: 'default' },
        () => {
          setDb(db)
          resolve()
        },
        reject,
      )
    })
  }

  run(sql: string, params: any): Promise<any> {
    return new Promise((resolve, reject) => {
      this.db.transaction((tx: any) => {
        tx.executeSql(sql, params || [], (t, result) => {
          //todo reject
          resolve()
        })
      })
    })
  }

  rows(sql: string, params: any): Promise<any> {
    return new Promise((resolve, reject) => {
      this.db.transaction((tx: any) => {
        tx.executeSql(sql, params || [], (t, result) => {
          //todo reject
          resolve(result.rows.raw())
        })
      })
    })
  }
}
