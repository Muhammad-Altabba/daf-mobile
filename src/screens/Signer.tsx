/**
 * Serto Mobile App
 *
 */

import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View, FlatList, TextInput } from 'react-native'
import { Query, Mutation, MutationState } from 'react-apollo'
import {
  Did,
  getDidsQuery,
  createDidMutation,
  importSeedMutation,
  deleteSeedMutation,
} from '../lib/Signer'
import { Container, Button, Constants, Screen } from '@kancha/kancha-ui'

interface Props {}

export default (props: Props) => {
  const { t } = useTranslation()
  const [seed, setSeed] = useState('')
  return (
    <Screen safeArea={true}>
      <Container>
        <Mutation
          mutation={importSeedMutation}
          refetchQueries={['getDids']}
          // tslint:disable-next-line:no-console
          onError={(e: any) => console.log('Error: ', e)}
        >
          {(mutate: any, state: MutationState) => {
            return (
              <Container>
                <TextInput
                  style={styles.textInput}
                  placeholder={'Enter seed phrase'}
                  value={seed}
                  onChangeText={setSeed}
                />

                <Button
                  type={Constants.BrandOptions.Primary}
                  block={Constants.ButtonBlocks.Filled}
                  buttonText={t('Import Seed')}
                  onPress={() => {
                    mutate({
                      variables: {
                        seed,
                      },
                    })
                  }}
                  navButton
                />
                {state.error !== undefined && (
                  <Text>{state.error.message}</Text>
                )}
              </Container>
            )
          }}
        </Mutation>

        <Mutation mutation={createDidMutation} refetchQueries={['getDids']}>
          {(mutate: any) => (
            <Button
              type={Constants.BrandOptions.Primary}
              block={Constants.ButtonBlocks.Filled}
              buttonText={t('Create New Identity')}
              onPress={() => {
                mutate()
              }}
              navButton
            />
          )}
        </Mutation>
      </Container>

      <Query query={getDidsQuery}>
        {({
          data,
          loading,
          refetch,
        }: {
          data: { dids: Did[] }
          loading: boolean
          refetch: () => void
        }) => (
          <FlatList
            data={data.dids}
            renderItem={({ item }) => <ListItem item={item} />}
            keyExtractor={(item, index) => item.did}
            onRefresh={refetch}
            refreshing={loading}
          />
        )}
      </Query>
    </Screen>
  )
}

const ListItem = ({ item }: { item: Did }) => {
  return (
    <View style={[styles.item]}>
      <View style={styles.header}>
        <Text style={styles.category}>{item.did}</Text>
      </View>
      <Text style={styles.message}>{item.seed}</Text>
      <Mutation mutation={deleteSeedMutation} refetchQueries={['getDids']}>
        {(mutate: any) => (
          <Button
            type={Constants.BrandOptions.Warning}
            block={Constants.ButtonBlocks.Outlined}
            buttonText={'Delete Seed'}
            onPress={() => {
              mutate({
                variables: {
                  address: item.address,
                },
              })
            }}
            navButton
          />
        )}
      </Mutation>
    </View>
  )
}

const styles = StyleSheet.create({
  item: {
    marginLeft: 3,
    marginTop: 1,
    marginBottom: 1,
    borderBottomWidth: 1,
    borderColor: 'gray',
    padding: 10,
  },
  header: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  category: {
    fontSize: 12,
    color: 'gray',
    fontWeight: 'bold',
  },
  message: {
    fontSize: 15,
  },
  date: {
    fontSize: 12,
    marginBottom: 5,
    color: 'gray',
  },
  textInput: {
    margin: 15,
    borderWidth: 1,
    fontSize: 14,
    borderColor: 'gray',
    padding: 10,
  },
})
