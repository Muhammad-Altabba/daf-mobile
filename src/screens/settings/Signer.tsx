/**
 * Serto Mobile App
 *
 */

import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { FlatList, TextInput } from 'react-native'
import { Query, Mutation } from 'react-apollo'
import { NavigationStackScreenProps } from 'react-navigation-stack'
// import { useQuery, useApolloClient } from '@apollo/react-hooks'
import {
  Did,
  getDidsQuery,
  createDidMutation,
  importSeedMutation,
} from '../../lib/Signer'
import {
  Container,
  Button,
  Constants,
  Screen,
  ListItem,
  Text,
} from '@kancha/kancha-ui'
import { Colors } from '../../theme'

interface SignerProps extends NavigationStackScreenProps {}

const Signer: React.FC<SignerProps> = props => {
  const { t } = useTranslation()
  const [seed, setSeed] = useState('')
  // const client = useApolloClient()
  // const { data, loading, error } = useQuery(getDidsQuery)
  return (
    <Screen
      safeArea={true}
      footerComponent={
        <Container padding>
          <Mutation
            mutation={importSeedMutation}
            refetchQueries={['getDids']}
            // tslint:disable-next-line:no-console
            onError={(e: any) => console.log('Error: ', e)}
          >
            {(mutate: any, state: any) => {
              return (
                <Container paddingBottom>
                  <Button
                    fullWidth
                    disabled={!seed}
                    type={Constants.BrandOptions.Primary}
                    block={Constants.ButtonBlocks.Filled}
                    buttonText={t('Import Seed')}
                    onPress={() => {
                      mutate({
                        variables: {
                          seed,
                        },
                      })
                    }}
                    navButton
                  />
                  {state.error !== undefined && (
                    <Text>{state.error.message}</Text>
                  )}
                </Container>
              )
            }}
          </Mutation>
          <Mutation mutation={createDidMutation} refetchQueries={['getDids']}>
            {(mutate: any) => (
              <Button
                fullWidth
                type={Constants.BrandOptions.Primary}
                block={Constants.ButtonBlocks.Filled}
                buttonText={t('Create New Identity')}
                onPress={() => {
                  mutate()
                }}
                navButton
              />
            )}
          </Mutation>
        </Container>
      }
    >
      <Container flex={1}>
        <Query query={getDidsQuery}>
          {({
            data,
            loading,
            refetch,
          }: {
            data: { dids: Did[]; selectedDid: string }
            loading: boolean
            refetch: () => void
          }) => (
            <FlatList
              style={{ backgroundColor: Colors.LIGHTEST_GREY, flex: 1 }}
              data={data && data.dids}
              renderItem={({ item, index }) => (
                <ListItem
                  last={index === data.dids.length - 1}
                  selected={item.isSelected}
                  onPress={() => {
                    props.navigation.push('DidViewer', {
                      did: item.did,
                      address: item.address,
                      seed: item.seed,
                    })
                  }}
                >
                  {item.did.substring(0, 20) + '...'}
                </ListItem>
              )}
              keyExtractor={item => item.did}
              onRefresh={refetch}
              refreshing={loading}
              ListHeaderComponent={
                <Container background={Constants.BrandOptions.Primary}>
                  <TextInput
                    style={{
                      borderWidth: 1,
                      padding: 15,
                      height: 100,
                      borderColor: Colors.LIGHT_GREY,
                    }}
                    placeholder={'Enter seed phrase'}
                    value={seed}
                    onChangeText={setSeed}
                  />
                </Container>
              }
            />
          )}
        </Query>
      </Container>
    </Screen>
  )
}

export default Signer
