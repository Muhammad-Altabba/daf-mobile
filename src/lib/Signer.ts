import { Resolvers, ApolloError } from 'apollo-client'
import gql from 'graphql-tag'
import { client } from './GraphQL'
import { RNUportHDSigner } from 'react-native-uport-signer'
import analytics from '@segment/analytics-react-native'
import Log from './Log'

// ANDROID OPTIONS
//
// `singleprompt` - Prompt is only asked once per session or period of time
// `prompt`       - Prompt every time
// `simple`       - Not hardware protected but you don't loose your key if you change pin
// `cloud`        - Backed up in some cloud storage

export const DEFAULT_LEVEL = 'simple'
export const SHOW_SEED_PROMPT = 'Do you want to reveal seed phrase?'

export interface Did {
  did: string
  address: string
  seed?: string
  __typename?: string
}

export const getDidsQuery = gql`
  query getDids {
    dids @client {
      did
      address
      seed
    }
  }
`
export const resolvers: Resolvers = {
  Did: {
    seed: async (did: Did, args, context) => {
      return await RNUportHDSigner.showSeed(did.address, SHOW_SEED_PROMPT)
    },
  },
  Query: {
    dids: async (_, args, context) => {
      const list = await RNUportHDSigner.listSeedAddresses()
      return list.map((address: string) => ({
        did: 'did:ethr:' + address,
        address,
        __typename: 'Did',
      })) as Did[]
    },
  },
  Mutation: {
    createDid: async (_, args, context) => {
      const { address } = await RNUportHDSigner.createSeed(DEFAULT_LEVEL)
      const did = 'did:ethr:' + address
      Log.info('Created: ' + did, 'Signer')
      analytics.track('Identity created', { type: 'did:ethr' })
      return {
        did,
        address,
        __typename: 'Did',
      } as Did
    },
    importSeed: async (_, { seed }, context) => {
      const { address } = await RNUportHDSigner.importSeed(seed, DEFAULT_LEVEL)
      const did = 'did:ethr:' + address
      Log.info('Seed imported: ' + did, 'Signer')
      analytics.track('Seed imported')
      return {
        did,
        address,
        __typename: 'Did',
      } as Did
    },
    deleteSeed: async (_, { address }, context) => {
      const result = await RNUportHDSigner.deleteSeed(address)
      const did = 'did:ethr:' + address
      Log.info('Deleted: ' + did, 'Signer')
      analytics.track('Identity deleted', { type: 'did:ethr' })
      return result
    },
  },
}

client.addResolvers(resolvers)

export const importSeedMutation = gql`
  mutation importSeed($seed: String!) {
    importSeed(seed: $seed) @client
  }
`

export const deleteSeedMutation = gql`
  mutation deleteSeed($address: String!) {
    deleteSeed(address: $address) @client
  }
`

export const createDidMutation = gql`
  mutation createDid {
    createDid @client {
      did
    }
  }
`
