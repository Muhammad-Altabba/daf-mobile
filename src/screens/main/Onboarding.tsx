import * as React from 'react'
import { useTranslation } from 'react-i18next'
import { Screen, Container, Text, Button, Constants } from '@kancha/kancha-ui'
import { NavigationScreen } from '../../navigators'
import { Query, Mutation } from 'react-apollo'
import { Did, getDidsQuery, createDidMutation } from '../../lib/Signer'
import { ActivityIndicator } from 'react-native'
import { Colors } from '../../theme'

interface OnboardingProps extends NavigationScreen {}

interface Resp {
  data: { dids: Did[] }
  loading: boolean
  refetch: () => void
  client: any
}

const Onboarding: React.FC<OnboardingProps> = props => {
  const { t } = useTranslation()

  return (
    <Query
      query={getDidsQuery}
      onCompleted={(data: { dids: Did[] }) => {
        if (data.dids.length > 0) {
          props.navigation.navigate('App')
        }
      }}
    >
      {({ data, loading, client }: Resp) => (
        <Screen>
          <Container alignItems={'center'} justifyContent={'center'} flex={1}>
            {!loading && data.dids.length === 0 ? (
              <Mutation
                mutation={createDidMutation}
                refetchQueries={['getDids']}
                onCompleted={({
                  createDid,
                }: {
                  createDid: {
                    address: string
                    did: string
                  }
                }) => {
                  // tslint:disable-next-line:no-console
                  console.log(createDid.address, createDid.did)

                  client.writeData({ data: { selectedDid: createDid.did } })
                }}
              >
                {(mutate: any) => (
                  <Button
                    fullWidth
                    type={Constants.BrandOptions.Primary}
                    block={Constants.ButtonBlocks.Filled}
                    buttonText={t('Create New Identity')}
                    onPress={() => {
                      mutate()
                    }}
                    navButton
                  />
                )}
              </Mutation>
            ) : (
              <ActivityIndicator size={'large'} color={Colors.BRAND} />
            )}
          </Container>
        </Screen>
      )}
    </Query>
  )
}

export default Onboarding
