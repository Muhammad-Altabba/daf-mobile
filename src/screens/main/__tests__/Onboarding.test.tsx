import 'react-native'
import React from 'react'
import { render } from 'react-native-testing-library'
import { MockedProvider } from '@apollo/react-testing'
import Onboarding from '../Onboarding'
import { Did, getDidsQuery, createDidMutation } from '../../../lib/Signer'

jest.useFakeTimers()

const mockDidItem: Did = {
  did: 'did:ethr:123456',
  address: '123456',
  isSelected: true,
  seed: 'winkle berry boojam',
}

const mocks = [
  {
    request: {
      query: getDidsQuery,
      variables: {},
    },
    result: {
      data: {
        logs: [mockDidItem],
      },
    },
  },
]

it('renders correctly', () => {
  const tree = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      {
        // @ts-ignore
        <Onboarding navigation={jest.fn()} />
      }
    </MockedProvider>,
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
