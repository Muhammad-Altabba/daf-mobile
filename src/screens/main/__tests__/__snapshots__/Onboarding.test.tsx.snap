// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": 1,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": 1,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": "center",
        "margin": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#3A89C9"
      hidesWhenStopped={true}
      size="large"
    />
  </View>
</View>
`;
